### PR TITLE
feat(pdfjs): upgrade to v4

### PR DESCRIPTION
**Pull request by Jordan-Hall**
Fixes: #1093

### CODE REVIEW

**Comment by shamoon** [comment]
Oh nice, I had thought that v4 would require substantive changes to the codebase but it seems the v3 changes were enough. Testing this locally things seem to work fine for me—great work.

**Comment by shamoon** [comment]
Ps. Also closes #1078

**Comment by Jordan-Hall** [comment]
> Oh nice, I had thought that v4 would require substantive changes to the codebase but it seems the v3 changes were enough. Testing this locally things seem to work fine for me—great work.
> 
> Ps. this also closes #1078

Yea it was thank you so much. Want to use this but hard time passing it through security :( even with your partial fix

**Comment by SimonFischer04** [comment]
Was just about to start working on the v4 port, but then saw this pr.
Tested this locally and everything works fine for me also.

Thanks for the great work @Jordan-Hall and @shamoon.

**Comment by Foxandxss** [comment]
Any updates on this? We have complains about this vulnerability and I am trying to hold it off but we might need to either fork or replace this library.

Thanks for the work tho!

**Comment by DerekLiang** [comment]
Could someone please approve the PR and make a release? Thank you!

**Comment by andrey-voytenko** [comment]
please process a PR

File: src/app/pdf-viewer/pdf-viewer.component.ts [hunk 1]
```diff
@@ -32,10 +32,11 @@ import type {
   PDFViewerOptions,
   ZoomScale
 } from './typings';
-import { PDFSinglePageViewer } from 'pdfjs-dist/web/pdf_viewer';
+import { PDFSinglePageViewer } from 'pdfjs-dist/web/pdf_viewer.mjs';
+import { GlobalWorkerOptions, VerbosityLevel, getDocument } from 'pdfjs-dist';
```
Comment (joh-klein) [inline src/app/pdf-viewer/pdf-viewer.component.ts:36]: Why do you import "pdfjs-dist" again when it is already imported in line 20?

Comment (Jordan-Hall) [reply src/app/pdf-viewer/pdf-viewer.component.ts:36]: Because its a deep import and its not available from the top level

Comment (joh-klein) [reply src/app/pdf-viewer/pdf-viewer.component.ts:36]: I might be a little rusty with my angular, but I always thought that this

```
import * as PDFJS from 'pdfjs-dist';

console.log(PDFJS.VerbosityLevel.INFOS)
```

is functionally the same as this.

```
import { VerbosityLevel } from 'pdfjs-dist';

console.log(VerbosityLevel.INFOS)
```

Sure - doing a deep import is more treeshakeable, and is thus preferable, but what you are telling me (or at least that is what I am understanding) is, that there is a difference and both imports are necessary. Did I get that right?

**Comment by jlbales** [comment]
Do we know how much longer until this is released? We may have to go with another library if it is not resolved this week.

File: src/app/pdf-viewer/pdf-viewer.component.ts (hunk 1 above)
Comment (DerekLiang) [reply src/app/pdf-viewer/pdf-viewer.component.ts:36]: Is this a show stopper?

Comment (Jordan-Hall) [reply src/app/pdf-viewer/pdf-viewer.component.ts:36]: No because  it has to be deep

**Comment by Jordan-Hall** [comment]
If its dark i'll release a fork or @shamoon might want too

**Comment by shamoon** [comment]
I think the author just updates it infrequently but I dunno. I don’t plan to create a fork.

**Comment by VadimDez** [comment]
I see few issues with this PR. One issue that I can reproduce is whenever I can toggle "Show all pages" - the same pdf will be appended on top of already rendered pdf.

@shamoon thank you so much for taking a look and trying to manage PRs/Issues!

**Comment by uright** [comment]
Thank you for @VadimDez for looking into this. Looking forward for the upgrade to take place. The #1109 dependency on pdfjs is preventing two of my applications to be released.

**Comment by Jordan-Hall** [comment]
> I see few issues with this PR. One issue that I can reproduce is whenever I can toggle "Show all pages" - the same pdf will be appended on top of already rendered pdf.
> 
> @shamoon thank you so much for taking a look and trying to manage PRs/Issues!

Hmm ok i'll take a look at that sorry

**Review by shamoon — COMMENTED** [review]
This seems to fix the show all pages issue for me (consolidating the import thing isnt important obv but it looks nice to me =):

```
--- a/src/app/pdf-viewer/pdf-viewer.component.ts
+++ b/src/app/pdf-viewer/pdf-viewer.component.ts
@@ -32,7 +32,6 @@ import type {
   PDFViewerOptions,
   ZoomScale
 } from './typings';
-import { PDFSinglePageViewer } from 'pdfjs-dist/web/pdf_viewer.mjs';
 import { GlobalWorkerOptions, VerbosityLevel, getDocument } from 'pdfjs-dist';

 if (!isSSR()) {
@@ -64,7 +63,7 @@ export class PdfViewerComponent
   public eventBus!: PDFJSViewer.EventBus;
   public pdfLinkService!: PDFJSViewer.PDFLinkService;
   public pdfFindController!: PDFJSViewer.PDFFindController;
-  public pdfViewer!: PDFJSViewer.PDFViewer | PDFSinglePageViewer;
+  public pdfViewer!: PDFJSViewer.PDFViewer | PDFJSViewer.PDFSinglePageViewer;

   private isVisible = false;

@@ -435,6 +434,10 @@ export class PdfViewerComponent
   }

   private setupViewer() {
+    if (this.pdfViewer) {
+      this.resetPdfDocument();
+    }
+
     assign(PDFJS, 'disableTextLayer', !this._renderText);

     this.initPDFServices();
```

**Comment by shamoon** [comment]
Actually I may have spoken to soon, it does fix rendering 1 page but I still see a duplicate when toggling. I can play with it again another time (unless you sort it in the meantime of course =)

**Comment by shamoon** [comment]
I think this should do it

```
--- a/src/app/pdf-viewer/pdf-viewer.component.ts
+++ b/src/app/pdf-viewer/pdf-viewer.component.ts
@@ -32,7 +32,6 @@ import type {
   PDFViewerOptions,
   ZoomScale
 } from './typings';
-import { PDFSinglePageViewer } from 'pdfjs-dist/web/pdf_viewer.mjs';
 import { GlobalWorkerOptions, VerbosityLevel, getDocument } from 'pdfjs-dist';

 if (!isSSR()) {
@@ -64,7 +63,7 @@ export class PdfViewerComponent
   public eventBus!: PDFJSViewer.EventBus;
   public pdfLinkService!: PDFJSViewer.PDFLinkService;
   public pdfFindController!: PDFJSViewer.PDFFindController;
-  public pdfViewer!: PDFJSViewer.PDFViewer | PDFSinglePageViewer;
+  public pdfViewer!: PDFJSViewer.PDFViewer | PDFJSViewer.PDFSinglePageViewer;

   private isVisible = false;

@@ -435,6 +434,10 @@ export class PdfViewerComponent
   }

   private setupViewer() {
+    if (this.pdfViewer) {
+      this.pdfViewer.setDocument(null as any);
+    }
+
     assign(PDFJS, 'disableTextLayer', !this._renderText);

     this.initPDFServices();
```

**Comment by csi-cdavis** [comment]
The README should probably be updated to note that Angular 17+ is required for the latest PDFJS version.

**Comment by shakoorattari** [comment]
When it will be released in npm?

**Comment by DerekLiang** [comment]
Can we publish a beta version of the NPM while we polishing the PR? Thank you!

**Comment by Bengejd** [comment]
@VadimDez any reason why this can't be merged?

**Comment by joh-klein** [comment]
Well – since the [issues noted](https://github.com/VadimDez/ng2-pdf-viewer/pull/1105#issuecomment-2282662538) by @VadimDez aren't fixed (even though @shamoon already[ wrote how to fix them](https://github.com/VadimDez/ng2-pdf-viewer/pull/1105#issuecomment-2282662538)), this will not move forward.

@Jordan-Hall can you please update the PR with the changes so this even has the chance of being looked at and merged?

**Comment by Bengejd** [comment]
> Well – since the [issues noted](https://github.com/VadimDez/ng2-pdf-viewer/pull/1105#issuecomment-2282662538) by @VadimDez aren't fixed (even though @shamoon already[ wrote how to fix them](https://github.com/VadimDez/ng2-pdf-viewer/pull/1105#issuecomment-2283276867)), this will not move forward.
> 
> @Jordan-Hall can you please update the PR with the changes so this even has the chance of being looked at and merged?

Since @Jordan-Hall hasn't responded in a while, I decided to resurrect the PR myself & address the issues brought forth. #1128 should satisfy those issues.

**Comment by Jordan-Hall** [comment]
It was on my list today. Sorry I no longer work for the company that needed this but was doing it today. But thank you so much for sorting this. I know I need it soon for personal projec

**Pull request closed without merging**